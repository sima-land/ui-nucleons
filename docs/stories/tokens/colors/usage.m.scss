@use 'pkg:@sima-land/ui-nucleons/colors';

.imports {
  display: flex;
  flex-direction: column;
  gap: 16px;
  margin-bottom: 32px;
  > * {
    width: 100%;
  }
}
