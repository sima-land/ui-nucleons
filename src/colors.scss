// GENERATED FILE - DO NOT CHANGE IT MANUALLY

// basic
$basic-blue: #1f84db;
$basic-gray87: #212121;
$basic-gray76: #3a3a3b;
$basic-gray66: #545455;
$basic-gray54: #757575;
$basic-gray38: #9e9e9e;
$basic-gray24: #c2c2c2;
$basic-gray12: #e0e0e0;
$basic-gray8: #ebebeb;
$basic-gray4: #f5f5f5;
$basic-gray2: #fafafa;
$basic-white: #fff;

// additional
$additional-deep-red: #d50000;
$additional-red: #fb3a2f;
$additional-light-red: #feebea;
$additional-dark-teal: #089176;
$additional-teal: #09ab8b;
$additional-green: #00c853;
$additional-light-green: #64dd17;
$additional-lime: #aeea00;
$additional-faded-green: #eff9ea;
$additional-pink: #e82e5c;
$additional-purple: #b52ea8;
$additional-violet: #902bd0;
$additional-deep-purple: #634bdf;
$additional-electric-blue: #2962ff;
$additional-light-blue: #0091ea;
$additional-cyan: #00b8d4;
$additional-sky: #e4f1f9;
$additional-deep-orange: #ff7200;
$additional-amber: #ffab00;
$additional-yellow: #ffd600;
$additional-gold: #d5a43b;
$additional-brown: #795548;
$additional-blue-gray: #607d8b;
$additional-deep-blue: #00599d;
$additional-dark-blue: #002b41;
$additional-unlit-blue: #1b75c2;
$additional-crimson: #f4446b;
