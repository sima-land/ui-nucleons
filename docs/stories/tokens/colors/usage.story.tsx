import { COLORS } from '@sima-land/ui-nucleons/colors';
import { useEffect, useRef, useState } from 'react';
import ClipboardJS from 'clipboard';
import { Button } from '@sima-land/ui-nucleons/button';
import { Layout } from '@sima-land/ui-nucleons/layout';
import { Input } from '@sima-land/ui-nucleons/input';
import { TextButton } from '@sima-land/ui-nucleons/text-button';
import CopySVG from '@sima-land/ui-quarks/icons/16x16/Stroked/Copy';
import CopyBigSVG from '@sima-land/ui-quarks/icons/24x24/Stroked/Copy';
import styles from './usage.m.scss';

export const meta = {
  category: 'Токены/Colors',
  title: 'Использование',
  parameters: {
    layout: 'padded',
  },
};

export default function Usage() {
  return (
    <Layout>
      <h2>Токены цветов</h2>

      <p>Библиотека предоставляет утилиты для использования цветов дизайн-системы.</p>

      <div className={styles.section}>
        <div className={styles.imports}>
          <ImportField label='Импорт SCSS' value="@use '@sima-land/ui-nucleons/colors';" />
          <ImportField
            label='Импорт JS'
            value="import { COLORS } from '@sima-land/ui-nucleons/colors';"
          />
        </div>
      </div>

      <div className={styles.section}>
        <div className={styles.colors}>
          {[...COLORS.entries()].map(([name, value], index) => (
            <ColorView key={index} {...{ name, value }} />
          ))}
        </div>
      </div>
    </Layout>
  );
}

function ImportField({ value, label }: { value: string; label: string }) {
  const inputRef = useRef<HTMLInputElement>(null);
  const buttonRef = useRef<HTMLButtonElement>(null);
  const [copied, setCopied] = useState(false);

  useEffect(() => {
    if (buttonRef.current) {
      const clipboard = new ClipboardJS(buttonRef.current, { text: () => value });

      clipboard.on('success', () => {
        setCopied(true);
        setTimeout(() => setCopied(false), 2000);
      });

      return () => clipboard.destroy();
    }
  }, []);

  return (
    <Input
      label={label}
      value={copied ? 'Скопировано' : value}
      disabled={copied}
      inputRef={inputRef}
      readOnly
      adornmentEnd={
        <TextButton
          disabled={copied}
          color='basic-gray87'
          buttonRef={buttonRef}
          onMouseDown={e => e.preventDefault()}
        >
          <CopyBigSVG fill='currentColor' />
        </TextButton>
      }
    />
  );
}

function ColorView({ name, value }: { name: string; value: string }) {
  return (
    <div className={styles.color}>
      <span className={styles.circle} style={{ background: value }} />
      <span className={styles.title}>{name}</span>
      <span>{value}</span>

      <CopyButton title='SCSS' value={`colors.$${name}`} />
      <CopyButton title='JS' value={`COLORS.get('${name}')`} />
    </div>
  );
}

function CopyButton({ title, value }: { title: string; value: string }) {
  const buttonRef = useRef<HTMLButtonElement>(null);
  const [copied, setCopied] = useState(false);

  useEffect(() => {
    if (buttonRef.current) {
      const clipboard = new ClipboardJS(buttonRef.current, {
        text: () => value,
      });

      clipboard.on('success', () => {
        setCopied(true);
        setTimeout(() => setCopied(false), 2000);
      });

      return () => clipboard.destroy();
    }
  }, []);

  return (
    <div>
      <Button
        ref={buttonRef}
        size='xs'
        viewType='secondary'
        icon={!copied ? CopySVG : undefined}
        disabled={copied}
      >
        {copied ? 'Скопировано' : title}
      </Button>
    </div>
  );
}
