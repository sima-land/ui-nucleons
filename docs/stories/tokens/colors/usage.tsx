import { COLORS } from '@sima-land/ui-nucleons/colors';
import { CardGrid, ImportField, TokenCard } from '../gradients/usage';
import styles from './usage.m.scss';

export function ColorsPicker() {
  return (
    <>
      <h2>Цвета</h2>

      <p>Библиотека предоставляет утилиты для использования цветов дизайн-системы.</p>

      <div className={styles.imports}>
        <ImportField label='Импорт SCSS' value="@use 'pkg:@sima-land/ui-nucleons/colors';" />
        <ImportField
          label='Импорт JS'
          value="import { COLORS } from '@sima-land/ui-nucleons/colors';"
        />
      </div>

      <CardGrid>
        {[...COLORS.entries()].map(([name, value], index) => (
          <TokenCard
            key={index}
            name={name}
            value={value}
            usageToken={`color/${name}`}
            usageScss={`colors.$${name}`}
            usageJs={`COLORS.get('${name}')`}
          />
        ))}
      </CardGrid>
    </>
  );
}
