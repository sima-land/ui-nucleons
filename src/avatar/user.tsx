import { useEffect, useState } from 'react';
import { sample } from 'lodash';
import { Avatar, AvatarProps } from '.';
import { Token } from '../colors';

export type UserAvatarProps = Omit<AvatarProps, 'bgColor' | 'bgOpacity' | 'textColor'>;

const COLORS: ReadonlyArray<Token> = [
  'additional-deep-red',
  'additional-red',
  'additional-teal',
  'additional-green',
  'additional-light-green',
  'additional-amber',
  'additional-purple',
  'additional-violet',
  'additional-deep-purple',
  'additional-electric-blue',
  'additional-light-blue',
  'additional-cyan',
  'additional-blue-gray',
  'additional-deep-orange',
];

export const colorKey = Symbol.for('user-avatar:color-key');

/**
 * Компонент аватара пользователя (не сотрудника).
 * @param props Свойства. Поддерживаются свойства Avatar.
 * @return Элемент.
 */
export function UserAvatar(props: UserAvatarProps) {
  const [color, setColor] = useState<(typeof COLORS)[number]>();

  useEffect(() => {
    const source: any = window;

    // запоминаем цвет (пока только один для всех аватаров на странице)
    source[colorKey] = source[colorKey] || sample(COLORS);

    setColor(source[colorKey]);
  }, []);

  return <Avatar {...props} bgColor={color} bgOpacity={color ? 0.48 : 0} textColor='basic-white' />;
}
