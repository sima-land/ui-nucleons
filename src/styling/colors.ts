import { ClassGetter } from './utils';
import { COLORS, Token } from '../colors';
import classes from './colors.module.scss';

const isColorKey = COLORS.has.bind(COLORS);

/** @deprecated */
export const color = ClassGetter<Token>(classes, isColorKey, 'color__');

/** @deprecated */
export const hoverColor = ClassGetter<Token>(classes, isColorKey, 'hover-color__');

/** @deprecated */
export const bgColor = ClassGetter<Token>(classes, isColorKey, 'bg-color__');

/** @deprecated */
export const stroke = ClassGetter<Token>(classes, isColorKey, 'stroke__');

/** @deprecated */
export const fill = ClassGetter<Token>(classes, isColorKey, 'fill__');
