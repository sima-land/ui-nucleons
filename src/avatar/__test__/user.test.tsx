import { mount } from 'enzyme';

import { UserAvatar, colorKey } from '../user';
import { COLORS } from '../../colors';

jest.mock('lodash', () => {
  const original = jest.requireActual('lodash');

  return {
    ...original,
    sample: jest.fn(arr => (Array.isArray(arr) ? arr[0] : undefined)),
  };
});

describe('UserAvatar', () => {
  beforeEach(() => {
    (window as any)[colorKey] = undefined;
  });

  it('should renders without props', () => {
    const wrapper = mount(<UserAvatar />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle props', () => {
    const wrapper1 = mount(
      <UserAvatar size={40} title='Hello World' style={{ clipPath: 'url(#fake-id)' }} />,
    );

    expect(wrapper1).toMatchSnapshot();
    expect((wrapper1.find('.root').prop('style') as any)['--avatar-color']).toEqual(
      COLORS.get((window as any)[colorKey]),
    );

    const wrapper2 = mount(
      <UserAvatar size={40} title='Hello World' style={{ clipPath: 'url(#fake-id)' }} />,
    );

    expect(wrapper2).toMatchSnapshot();
    expect((wrapper2.find('.root').prop('style') as any)['--avatar-color']).toEqual(
      COLORS.get((window as any)[colorKey]),
    );
  });
});
