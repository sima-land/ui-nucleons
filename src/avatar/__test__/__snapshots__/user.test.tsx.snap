// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserAvatar should handle props 1`] = `
<UserAvatar
  size={40}
  style={
    {
      "clipPath": "url(#fake-id)",
    }
  }
  title="Hello World"
>
  <Avatar
    bgColor="additional-deep-red"
    bgOpacity={0.48}
    size={40}
    style={
      {
        "clipPath": "url(#fake-id)",
      }
    }
    textColor="basic-white"
    title="Hello World"
  >
    <div
      className="root size-40 color__basic-white"
      data-testid="avatar"
      style={
        {
          "--avatar-color": "#d50000",
          "--avatar-color-opacity": 0.48,
          "clipPath": "url(#fake-id)",
        }
      }
    >
      <span
        className="monogram"
      >
        WH
      </span>
    </div>
  </Avatar>
</UserAvatar>
`;

exports[`UserAvatar should handle props 2`] = `
<UserAvatar
  size={40}
  style={
    {
      "clipPath": "url(#fake-id)",
    }
  }
  title="Hello World"
>
  <Avatar
    bgColor="additional-deep-red"
    bgOpacity={0.48}
    size={40}
    style={
      {
        "clipPath": "url(#fake-id)",
      }
    }
    textColor="basic-white"
    title="Hello World"
  >
    <div
      className="root size-40 color__basic-white"
      data-testid="avatar"
      style={
        {
          "--avatar-color": "#d50000",
          "--avatar-color-opacity": 0.48,
          "clipPath": "url(#fake-id)",
        }
      }
    >
      <span
        className="monogram"
      >
        WH
      </span>
    </div>
  </Avatar>
</UserAvatar>
`;

exports[`UserAvatar should renders without props 1`] = `
<UserAvatar>
  <Avatar
    bgColor="additional-deep-red"
    bgOpacity={0.48}
    textColor="basic-white"
  >
    <div
      className="root size-72 color__basic-white"
      data-testid="avatar"
      style={
        {
          "--avatar-color": "#d50000",
          "--avatar-color-opacity": 0.48,
        }
      }
    >
      <ForwardRef(PersonSVG)
        className="icon"
        fill="#fff"
        height={24}
        width={24}
      >
        <svg
          className="icon"
          fill="#fff"
          height={24}
          viewBox="0 0 24 24"
          width={24}
        >
          <path
            d="M6.936 14.149a1 1 0 01-.585 1.287c-1.69.634-2.512 1.774-2.93 2.765a5.66 5.66 0 00-.37 1.305 4.584 4.584 0 00-.05.482V20.009l.005.085c.007.08.02.19.048.315.055.248.142.451.248.578A1.088 1.088 0 003.47 21H20.53a1.088 1.088 0 00.168-.013c.106-.127.193-.33.248-.578a2.473 2.473 0 00.053-.4V19.988a4.575 4.575 0 00-.05-.481 5.66 5.66 0 00-.37-1.306c-.418-.99-1.24-2.13-2.93-2.765a1 1 0 01.702-1.872c2.31.866 3.488 2.476 4.07 3.86.289.685.432 1.313.504 1.773a6.581 6.581 0 01.073.725l.002.05V20L22 20l1 .001v.03a2.904 2.904 0 01-.013.236 4.49 4.49 0 01-.088.573c-.095.431-.306 1.074-.806 1.584-.51.518-1.181.576-1.563.576H3.47c-.382 0-1.053-.058-1.562-.576-.501-.51-.712-1.153-.807-1.584a4.474 4.474 0 01-.1-.809v-.019L1 20.005v-.003S1 20 2 20l-1-.001V19.972a3.002 3.002 0 01.01-.216 7.662 7.662 0 01.568-2.331c.583-1.385 1.761-2.995 4.07-3.861a1 1 0 011.288.585z"
          />
          <path
            clipRule="evenodd"
            d="M12 3C9.878 3 8 4.923 8 7.5S9.878 12 12 12s4-1.923 4-4.5S14.122 3 12 3zM6 7.5C6 4.002 8.599 1 12 1s6 3.002 6 6.5-2.599 6.5-6 6.5-6-3.002-6-6.5z"
            fillRule="evenodd"
          />
        </svg>
      </ForwardRef(PersonSVG)>
    </div>
  </Avatar>
</UserAvatar>
`;
